pkg/migrate/cstor/blockdevice_correction.go

```diff
@@ -0,0 +1,361 @@
+		}
+	}
+
+	newCSPObj := cspObj.DeepCopy()
+	klog.Infof("Correcting the following bd in csp %s \n %v", cspObj.Name, correctCSPBD)
+	for i, rg := range newCSPObj.Spec.Group {
+		for j, bd := range rg.Item {
+			if correctCSPBD[bd.Name] != "" {
+				err = c.updateBDRefsAndlabels(spcObj, bd.Name, correctCSPBD[bd.Name])
+				if err != nil {
+					return errors.Wrapf(err, "failed to update old bd %s & new bd %s for spc %s",
+						bd.Name, correctCSPBD[bd.Name], spcObj.Name,
+					)
+				}
+				newCSPObj.Spec.Group[i].Item[j].Name = correctCSPBD[bd.Name]
+				// this field will automatically corrected by the operator
+				newCSPObj.Spec.Group[i].Item[j].DeviceID = ""
+			}
+		}
+	}
+	data, err := GetPatchData(cspObj, newCSPObj)
+	if err != nil {
+		return err
+	}
+	_, err = csp.KubeClient().
+		Patch(cspObj.Name, types.MergePatchType, data)
+
+	return err
+}
+
+func (c *CSPCMigrator) findBDforDevlink(devlink, hostname string) (string, error) {
+	bds, err := c.OpenebsClientset.OpenebsV1alpha1().
+		BlockDevices(c.OpenebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: "kubernetes.io/hostname=" + hostname,
+		})
+	if err != nil {
+		return "", err
+	}
+	for _, bd := range bds.Items {
+		if bd.Spec.Path != "" {
+			if strings.Contains(devlink, bd.Spec.Path) {
+				ok, err := c.verifyBDStatus(bd, hostname)
+				if err != nil {
+					return "", err
+				}
+				if ok {
+					return bd.Name, nil
+				}
+			}
+		}
+		for _, devLinks := range bd.Spec.DevLinks {
+			for _, link := range devLinks.Links {
+				if strings.Contains(devlink, link) {
+					ok, err := c.verifyBDStatus(bd, hostname)
+					if err != nil {
+						return "", err
+					}
+					if ok {
+						return bd.Name, nil
+					}
+				}
+			}
+		}
+	}
+	return "", errors.Errorf("blockdevice not found for devlink %s", devlink)
+}
+
+func (c *CSPCMigrator) verifyBDStatus(bdObj v1alpha1.BlockDevice, hostName string) (bool, error) {
+	if bdObj.Status.State == v1alpha1.BlockDeviceActive {
+		nodes, err := c.KubeClientset.CoreV1().Nodes().
+			List(metav1.ListOptions{
+				LabelSelector: openebstypes.HostNameLabelKey + "=" + hostName,
+			})
+		if err != nil {
+			return false, err
+		}
+		if len(nodes.Items) == 1 {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (c *CSPCMigrator) updateBDRefsAndlabels(spcObj *apis.StoragePoolClaim, oldBD, newBD string) error {
+	spcKind := "StoragePoolClaim"
+	oldBDObj, err := c.OpenebsClientset.OpenebsV1alpha1().
+		BlockDevices(c.OpenebsNamespace).Get(oldBD, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	newBDObj, err := c.OpenebsClientset.OpenebsV1alpha1().
+		BlockDevices(c.OpenebsNamespace).Get(newBD, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// For the new bd
+	// - if bdc already present:
+	//   - if ownerref already present: update the bdc with correct ownerref & label
+	//   - if ownerref not present: add the spc ownerref & label bdc
+	if newBDObj.Spec.ClaimRef.Name != spcObj.Name {
```

Comment:
BD claimRef will be always holding the BDC, not the SPC. 
**Note**: ClaimRef is a pointer we have to check for nil before accessing it(new BD may not be claimed).